tests/unit/sessions/test_session_pool.py

```diff
@@ -162,3 +162,27 @@ async def test_session_pool_persist_and_restore(event_manager: EventManager, kvs
         await sp.reset_store()
         previous_state = await kvs.get_value(key=PERSIST_STATE_KEY)
         assert previous_state is None
+
+
+async def test_methods_raise_error_when_not_active() -> None:
+    session = Session()
+    session_pool = SessionPool()
+
+    assert session_pool.active is False
+
+    with pytest.raises(RuntimeError, match='SessionPool is not active.'):
+        session_pool.get_state(as_dict=True)
+
+    with pytest.raises(RuntimeError, match='SessionPool is not active.'):
+        session_pool.add_session(session)
+
+    with pytest.raises(RuntimeError, match='SessionPool is not active.'):
+        await session_pool.get_session()
+
+    with pytest.raises(RuntimeError, match='SessionPool is not active.'):
+        await session_pool.get_session_by_id(session.id)
+
+    await session_pool.reset_store()
+
+    async with session_pool:
+        assert session_pool.active is True
```

Comment:
I think that this is can lead to unexpected situations. Edit following tests like this to see. You will be still in outer context, but session_pool.active will be false. 
This implementation let's consumers of these context managers enter many times, but once single context has been left, all other context levels are no longer active. 

I think it is good idea to give consumers of those context managers "active" flag, but nothing else. Consumers should safeguard usage of such contexts themselves. When you are in the context, you do not want(users can decide) to enter it again, but when you are "twice" in the context and you leave the inner context, you still want(users can decide) to be in the outer context. 

```suggestion
    async with session_pool:
        async with session_pool:
            pass
        assert session_pool.active is True
        
```

Reply:
I think this is resolved as now we raise an exception in case of multiple enters/exits.